### PR TITLE
Circuit for opcode SHR

### PR DESCRIPTION
**Pull request by icemelon**
Original comments in #77

### CODE REVIEW

File: zkevm-circuits/src/evm_circuit/util/math_gadget.rs [hunk 1]
```diff
@@ -820,3 +823,396 @@ impl<F: FieldExt, const N_BYTES: usize> MinMaxGadget<F, N_BYTES> {
...
+    ) -> Self {
+        let b = cb.query_word();
+        let a_slice_front = array_init::array_init(|_| cb.query_byte());
+        let a_slice_back = array_init::array_init(|_| cb.query_byte());
+        let shift_div64 = cb.query_cell();
+        let shift_mod64_div8 = cb.query_cell();
+        let shift_mod64_decpow = cb.query_cell();
+        let shift_mod64_pow = cb.query_cell();
+        let shift_mod8 = cb.query_cell();
+        let shift_overflow = cb.query_bool();
```
Comment (icemelon) [inline zkevm-circuits/src/evm_circuit/util/math_gadget.rs:887]: No need to allocate a cell for `shift_overflow`.

File: zkevm-circuits/src/evm_circuit/util/math_gadget.rs [hunk 2]
```diff
@@ -820,3 +823,396 @@ impl<F: FieldExt, const N_BYTES: usize> MinMaxGadget<F, N_BYTES> {
...
+
+        // check (combination of shift[1..32] == 0) == 1 - shift_overflow
+        let mut sum = 0.expr();
+        (1..32).for_each(|idx| sum = sum.clone() + shift.cells[idx].expr());
+        let is_zero = IsZeroGadget::construct(cb, sum);
```
Comment (icemelon) [inline zkevm-circuits/src/evm_circuit/util/math_gadget.rs:892]: Add `let shift_overflow = 1.expr() - is_zero.expr();`

File: zkevm-circuits/src/evm_circuit/util/math_gadget.rs [hunk 3]
```diff
@@ -820,3 +823,396 @@ impl<F: FieldExt, const N_BYTES: usize> MinMaxGadget<F, N_BYTES> {
...
+}
+impl<F: FieldExt> ShrWordsGadget<F> {
+    pub(crate) fn construct(
+        cb: &mut ConstraintBuilder<F>,
+        a: util::Word<F>,
+        shift: util::Word<F>,
+    ) -> Self {
+        let b = cb.query_word();
+        let a_slice_front = array_init::array_init(|_| cb.query_byte());
+        let a_slice_back = array_init::array_init(|_| cb.query_byte());
```
Comment (icemelon) [inline zkevm-circuits/src/evm_circuit/util/math_gadget.rs:881]:
```suggestion
        let a_slice_lo = array_init::array_init(|_| cb.query_byte());
        let a_slice_hi = array_init::array_init(|_| cb.query_byte());
```

File: zkevm-circuits/src/evm_circuit/util/math_gadget.rs [hunk 4]
```diff
@@ -820,3 +823,396 @@ impl<F: FieldExt, const N_BYTES: usize> MinMaxGadget<F, N_BYTES> {
...
+    // slice_back means the lower part of the split digit
+    a_slice_front: [Cell<F>; 32],
+    a_slice_back: [Cell<F>; 32],
+    // shift_div64, shift_mod64_div8, shift_mod8
+    // is used to seperate shift[0]
+    shift_div64: Cell<F>,
+    shift_mod64_div8: Cell<F>,
+    shift_mod64_decpow: Cell<F>, // means 2^(8-shift_mod64)
+    shift_mod64_pow: Cell<F>,    // means 2^shift_mod64
+    shift_mod8: Cell<F>,
```
Comment (icemelon) [inline zkevm-circuits/src/evm_circuit/util/math_gadget.rs:866]: move this line to line 864

File: zkevm-circuits/src/evm_circuit/util/math_gadget.rs [hunk 5]
```diff
@@ -820,3 +823,396 @@ impl<F: FieldExt, const N_BYTES: usize> MinMaxGadget<F, N_BYTES> {
...
+
+#[derive(Clone, Debug)]
+pub struct ShrWordsGadget<F> {
+    a: util::Word<F>,
+    shift: util::Word<F>,
+    b: util::Word<F>,
+    // slice_front means the higher part of split digit
+    // slice_back means the lower part of the split digit
+    a_slice_front: [Cell<F>; 32],
+    a_slice_back: [Cell<F>; 32],
```
Comment (icemelon) [inline zkevm-circuits/src/evm_circuit/util/math_gadget.rs:859]:
```suggestion
    a_slice_hi: [Cell<F>; 32],
    a_slice_lo: [Cell<F>; 32],
```

Using lower / higher bits are better than front / back.

File: zkevm-circuits/src/evm_circuit/util/math_gadget.rs [hunk 6]
```diff
@@ -820,3 +823,396 @@ impl<F: FieldExt, const N_BYTES: usize> MinMaxGadget<F, N_BYTES> {
...
+        cb.require_equal(
+            "shift_overflow == shift > 256 ",
+            shift_overflow.expr(),
+            1.expr() - is_zero.expr(),
+        );
+
+        // rename variable:
+        // shift_div64 :a
+        // shift_mod64_div8:b
+        // shift_mod8:c
```
Comment (icemelon) [inline zkevm-circuits/src/evm_circuit/util/math_gadget.rs:902]: Remove these comments as it's no longer there.

File: zkevm-circuits/src/evm_circuit/util/math_gadget.rs [hunk 7]
```diff
@@ -820,3 +823,396 @@ impl<F: FieldExt, const N_BYTES: usize> MinMaxGadget<F, N_BYTES> {
...
+            shift.cells[0].expr(),
+            shift_div64.expr() * 64.expr() + shift_mod64.clone(),
+        );
+
+        // merge 8 8-bit cell for a 64-bit expression
+        // for a, a_slice_front, a_slice_back, b
+        let mut a_digits = vec![];
+        let mut a_slice_front_digits = vec![];
+        let mut a_slice_back_digits = vec![];
+        let mut b_digits = vec![];
```
Comment (icemelon) [inline zkevm-circuits/src/evm_circuit/util/math_gadget.rs:918]:
```suggestion
        let mut a_limb64 = vec![];
        let mut b_limb64 = vec![];
        let mut a_slice_lo_limb64 = vec![];
        let mut a_slice_hi_limb64 = vec![];
```

File: zkevm-circuits/src/evm_circuit/util/math_gadget.rs [hunk 8]
```diff
@@ -820,3 +823,396 @@ impl<F: FieldExt, const N_BYTES: usize> MinMaxGadget<F, N_BYTES> {
...
+            a_slice_front_digits
+                .push(from_bytes::expr(&a_slice_front[now_idx..now_idx + 8]));
+            b_digits.push(from_bytes::expr(&b.cells[now_idx..now_idx + 8]));
+        }
+
+        // check combination of a_slice_back_digits and a_slice_front_digits
+        // == b_digits
+        let mut shr_constraints =
+            (0..4).map(|_| 0.expr()).collect::<Vec<Expression<F>>>();
+        for transplacement in (0_usize)..(4_usize) {
```
Comment (icemelon) [inline zkevm-circuits/src/evm_circuit/util/math_gadget.rs:933]:
```suggestion
        for transplacement in 0..4 {
```

File: zkevm-circuits/src/evm_circuit/util/math_gadget.rs [hunk 9]
```diff
@@ -820,3 +823,396 @@ impl<F: FieldExt, const N_BYTES: usize> MinMaxGadget<F, N_BYTES> {
...
+// This function generates Lagrange polynomial given a cell, index, and domain
+// size. The polynomial will be equal to 1 when `cell == idx`, otherwise 0.
+// The value of the cell needs to be in the range [0, domain_size)
+fn generate_lagrange_base_polynomial<F: FieldExt>(
+    cell: Cell<F>,
+    idx: u64,
+    domain_size: u64,
+) -> Expression<F> {
+    let mut base_poly = 1.expr();
+    let mut accumulated_inverse = 1.expr();
```
Comment (icemelon) [inline zkevm-circuits/src/evm_circuit/util/math_gadget.rs:836]:
```suggestion
    let mut denominator = 1.expr();
```

File: zkevm-circuits/src/evm_circuit/util/math_gadget.rs [hunk 10]
```diff
@@ -820,3 +823,396 @@ impl<F: FieldExt, const N_BYTES: usize> MinMaxGadget<F, N_BYTES> {
...
+
+// This function generates Lagrange polynomial given a cell, index, and domain
+// size. The polynomial will be equal to 1 when `cell == idx`, otherwise 0.
+// The value of the cell needs to be in the range [0, domain_size)
+fn generate_lagrange_base_polynomial<F: FieldExt>(
+    cell: Cell<F>,
+    idx: u64,
+    domain_size: u64,
+) -> Expression<F> {
+    let mut base_poly = 1.expr();
```
Comment (icemelon) [inline zkevm-circuits/src/evm_circuit/util/math_gadget.rs:835]:
```suggestion
    let mut nominator = 1.expr();
```

File: zkevm-circuits/src/evm_circuit/util/math_gadget.rs [hunk 11]
```diff
@@ -820,3 +823,396 @@ impl<F: FieldExt, const N_BYTES: usize> MinMaxGadget<F, N_BYTES> {
...
+    let mut accumulated_inverse = 1.expr();
+    for x in 0..domain_size {
+        if x != idx {
+            base_poly = base_poly * (cell.expr() - x.expr());
+            let inverse = if x < idx {
+                F::from_u128((idx - x) as u128).invert().unwrap()
+            } else {
+                -F::from_u128((x - idx) as u128).invert().unwrap()
+            };
+            accumulated_inverse = accumulated_inverse * inverse;
```
Comment (icemelon) [inline zkevm-circuits/src/evm_circuit/util/math_gadget.rs:845]: You can first do the multiplication and do the inversion at last.

File: zkevm-circuits/src/evm_circuit/util/math_gadget.rs [hunk 12]
```diff
@@ -820,3 +823,396 @@ impl<F: FieldExt, const N_BYTES: usize> MinMaxGadget<F, N_BYTES> {
...
+        }
+    }
+    base_poly * accumulated_inverse
```
Comment (icemelon) [inline zkevm-circuits/src/evm_circuit/util/math_gadget.rs:848]:
```suggestion
    nominator * denominator.invert().unwrap()
```

File: zkevm-circuits/src/evm_circuit/util/math_gadget.rs [hunk 13]
```diff
@@ -820,3 +823,396 @@ impl<F: FieldExt, const N_BYTES: usize> MinMaxGadget<F, N_BYTES> {
...
+        }
+
+        // check combination of a_slice_back_digits and a_slice_front_digits
+        // == b_digits
+        let mut shr_constraints =
+            (0..4).map(|_| 0.expr()).collect::<Vec<Expression<F>>>();
+        for transplacement in (0_usize)..(4_usize) {
+            // generate the polynomial depends on the shift_div64
+            let select_transplacement_polynomial =
+                generate_lagrange_base_polynomial(
```
Comment (icemelon) [inline zkevm-circuits/src/evm_circuit/util/math_gadget.rs:936]: You can replace the lagrange poly by a lookup table as you only have 64 * 4 = 256 cases in total. So you can avoid the high degree expression.

**Review by icemelon — COMMENTED** [review]
Another idea to avoid the high degree lagrange polynomial is that instead of using 64-bit for `a_digits` and `b_digits`, you could probably use 128-bit. It seems that it won't exponentially increase the lookup table size.

File: zkevm-circuits/src/evm_circuit/util/math_gadget.rs (hunk 10 above)
Comment (z2trillion) [reply zkevm-circuits/src/evm_circuit/util/math_gadget.rs:798]: should be `numerator`

**Comment by lispc** [comment]
i suggest open another PR instead of this. It seems some github command is incorrect resulting every strange diffs and commits history...
image

My guess is you rebase main onto shr_new, which is not correct... Either (A) merge main into shr_new, or (B) rebase shr_new onto main (as a new branch) and force push

**Comment by lispc** [comment]
need to add: `gas_left: Delta(-OpcodeId::SHR.constant_gas_cost().expr()),`

**Comment by lispc** [comment]
https://github.com/scroll-tech/zkevm-circuits/pull/137

**Pull request closed without merging**